### ISSUE TITLE
Collab close and rollover problems on testnet

### ISSUE THREAD

**Issue by da-kami** [itchysats/itchysats#2199]
We ran into problem with CFD `aabb011f-9a22-40bb-b6cc-45c2c2a81fd7`.

## `taker_master` logs

```
2022-06-08 12:20:50	
{"timestamp":"2022-06-08T02:20:50.01898658Z","level":"DEBUG","fields":{"message":"Applying new events to CFD","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7","aggregate":"model::cfd::Cfd","cfd_version":"7","num_events":"0"},"target":"sqlite_db"}
	
2022-06-08 12:20:50	
{"timestamp":"2022-06-08T02:20:50.026144209Z","level":"INFO","fields":{"message":"Appended event to database","event":"CollaborativeSettlementStarted","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7"},"target":"sqlite_db"}
	
2022-06-08 12:20:50	
{"timestamp":"2022-06-08T02:20:50.026448835Z","level":"DEBUG","fields":{"message":"Handled request","method":"POST","path":"/api/cfd/aabb011f-9a22-40bb-b6cc-45c2c2a81fd7/settle","status":"200 OK"},"target":"http"}
	
2022-06-08 12:20:50	
{"timestamp":"2022-06-08T02:20:50.027357261Z","level":"DEBUG","fields":{"message":"Applying new events to CFD","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7","aggregate":"daemon::projection::Cfd","cfd_version":"7","num_events":"1"},"target":"sqlite_db"}
	
2022-06-08 12:20:50	
{"timestamp":"2022-06-08T02:20:50.035812097Z","level":"DEBUG","fields":{"message":"Applying new events to CFD","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7","aggregate":"daemon::position_metrics::Cfd","cfd_version":"7","num_events":"1"},"target":"sqlite_db"}
	
2022-06-08 12:20:53	
{"timestamp":"2022-06-08T02:20:53.140958891Z","level":"ERROR","fields":{"message":"Rollover failed: End of stream while receiving Decision","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7"},"target":"daemon::collab_settlement::protocol"}
	
2022-06-08 12:20:53	
{"timestamp":"2022-06-08T02:20:53.142207492Z","level":"DEBUG","fields":{"message":"Applying new events to CFD","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7","aggregate":"model::cfd::Cfd","cfd_version":"7","num_events":"1"},"target":"sqlite_db"}
	
2022-06-08 12:20:53	
{"timestamp":"2022-06-08T02:20:53.142764198Z","level":"WARN","fields":{"message":"Collaborative settlement failed: End of stream while receiving Decision","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7"},"target":"model::cfd"}
	
2022-06-08 12:20:53	
{"timestamp":"2022-06-08T02:20:53.151720153Z","level":"INFO","fields":{"message":"Appended event to database","event":"CollaborativeSettlementFailed","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7"},"target":"sqlite_db"}
```

And later rollover ran into trouble with this CFD:

```
2022-06-08 13:02:05	
{"timestamp":"2022-06-08T03:02:05.174389754Z","level":"ERROR","fields":{"message":"Rollover failed: End of stream while receiving rollover decision from maker","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7"},"target":"daemon::rollover::protocol"}
	
2022-06-08 13:02:05	
{"timestamp":"2022-06-08T03:02:05.175288564Z","level":"DEBUG","fields":{"message":"Applying new events to CFD","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7","aggregate":"model::cfd::Cfd","cfd_version":"9","num_events":"1"},"target":"sqlite_db"}
	
2022-06-08 13:02:05	
{"timestamp":"2022-06-08T03:02:05.175717612Z","level":"WARN","fields":{"message":"Rollover failed: End of stream while receiving rollover decision from maker","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7"},"target":"model::cfd"}Show context
	
2022-06-08 13:02:05	
{"timestamp":"2022-06-08T03:02:05.179112123Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverFailed","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7"},"target":"sqlite_db"}
```

and then:

```
2022-06-08 13:07:05	
{"timestamp":"2022-06-08T03:07:05.527961805Z","level":"INFO","fields":{"message":"Rollover proposal got accepted","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7"},"target":"daemon::rollover::taker"}Show context
	
2022-06-08 13:07:05	
{"timestamp":"2022-06-08T03:07:05.546760813Z","level":"DEBUG","fields":{"message":"Applying new events to CFD","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7","aggregate":"daemon::position_metrics::Cfd","cfd_version":"12","num_events":"1"},"target":"sqlite_db"}
	
2022-06-08 13:07:14	
{"timestamp":"2022-06-08T03:07:14.883172161Z","level":"ERROR","fields":{"message":"Rollover failed: Counterparty sent invalid revocation sk","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7"},"target":"daemon::rollover::protocol"}
	
2022-06-08 13:07:14	
{"timestamp":"2022-06-08T03:07:14.884128743Z","level":"DEBUG","fields":{"message":"Applying new events to CFD","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7","aggregate":"model::cfd::Cfd","cfd_version":"12","num_events":"1"},"target":"sqlite_db"}
	
2022-06-08 13:07:14	
{"timestamp":"2022-06-08T03:07:14.885425984Z","level":"WARN","fields":{"message":"Rollover failed: Counterparty sent invalid revocation sk","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7"},"target":"model::cfd"}
```

It kept failing with this problem on both sides and will be force closed.

## `maker` logs

Collab settlement failed to start error:

```
2022-06-08 12:20:50	
{"timestamp":"2022-06-08T02:20:50.130892495Z","level":"ERROR","fields":{"message":"Rollover failed: Failed to start collab settlement protocol: Failed to execute command on CFD: Proposed collab settlement does not equal locally created one. Local: Transaction { version: 2, lock_time: 0, input: [TxIn { previous_output: OutPoint { txid: f0ebf41aff66609349e80593fe514908477283e1f413904f0e5732c4c93b3feb, vout: 0 }, script_sig: Script(), sequence: 4294967295, witness: [] }], output: [TxOut { value: 328457, script_pubkey: Script(OP_0 OP_PUSHBYTES_20 ba6b63c217bf0ec2938b54ba04739a56d9bfeea0) }, TxOut { value: 158637, script_pubkey: Script(OP_0 OP_PUSHBYTES_20 416777184a21af22328634360c2a2dcfa8f6c3aa) }] }, proposed: Transaction { version: 2, lock_time: 0, input: [TxIn { previous_output: OutPoint { txid: f0ebf41aff66609349e80593fe514908477283e1f413904f0e5732c4c93b3feb, vout: 0 }, script_sig: Script(), sequence: 4294967295, witness: [] }], output: [TxOut { value: 328455, script_pubkey: Script(OP_0 OP_PUSHBYTES_20 ba6b63c217bf0ec2938b54ba04739a56d9bfeea0) }, TxOut { value: 158639, script_pubkey: Script(OP_0 OP_PUSHBYTES_20 416777184a21af22328634360c2a2dcfa8f6c3aa) }] }","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7"},"target":"daemon::collab_settlement::protocol"}
```

**Comment by da-kami** [comment]
## Analysis

### Problem of failing rollovers

The first rollover completed for the taker, but not for the maker, all consecutive rollovers fail because they disagree on the latest rollover.

There is only one `RolloverCompleted` event recorded in the logs of the taker for this CFD:

```
2022-06-08 11:02:15	
{"timestamp":"2022-06-08T01:02:15.756899998Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverCompleted","order_id":"aabb011f-9a22-40bb-b6cc-45c2c2a81fd7"},"target":"sqlite_db"}
```

There is no `RolloverCompleted` event ever recorded for this CFD on the maker.

The rollover on the maker side just ends after `RolloverAccepted` - there is not error / warning related to this CFD and this rollover in the logs.

However, I found this in the maker logs:

```
2022-06-08 11:02:15	
thread 'rocket-worker-thread' panicked at 'SelectNextSome polled after terminated', /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/stream/stream/select_next_some.rs:32:9
	
2022-06-08 11:02:15	
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

The timestamp matches the time when `RolloverCompleted` was recorded on the taker side.

I think that panic might have cause the maker to fail. I don't understand where this panic is coming from.
This is the only `panic` in the logs of the maker within the last 7 days. 
Unfortunately I cannot explain further what exactly failed on the maker that would cause this panic.

### Problem of failing collaborative settlement

Taker and maker disagree on the amounts in `TxOut` of the proposal transaction:

maker (i.e. `Local`:

```
output: [
    TxOut
    {
      value: 328457,
      script_pubkey: Script(OP_0
      OP_PUSHBYTES_20
      ba6b63c217bf0ec2938b54ba04739a56d9bfeea0)
    },
    TxOut
    {
      value: 158637,
      script_pubkey: Script(OP_0
      OP_PUSHBYTES_20
      416777184a21af22328634360c2a2dcfa8f6c3aa)
    }
  ]
```

taker (i.e. `proposed`):


```
output: [
    TxOut
    {
      value: 328455,
      script_pubkey: Script(OP_0
      OP_PUSHBYTES_20
      ba6b63c217bf0ec2938b54ba04739a56d9bfeea0)
    },
    TxOut
    {
      value: 158639,
      script_pubkey: Script(OP_0
      OP_PUSHBYTES_20
      416777184a21af22328634360c2a2dcfa8f6c3aa)
    }
  ]
```

This is because of the failed rollover, because the parties disagree on the fees now.
When we create a proposal the fees are taken into account:

https://github.com/itchysats/itchysats/blob/78d4fdef62a87fd3f3921f9f59b24bef74e6b157/model/src/cfd.rs#L1079-L1093

Since the taker is one rollover "ahead" the fee account is different.
So, the problematic rollover cause collaborative close to fail as well.

**Comment by da-kami** [comment]
This ticket boils down to understanding the panic

```
2022-06-08 11:02:15	
thread 'rocket-worker-thread' panicked at 'SelectNextSome polled after terminated', /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/stream/stream/select_next_some.rs:32:9
	
2022-06-08 11:02:15	
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Here is my theory:

The taker sent rollover `Msg3` first. It then received `Msg3` from the maker, finished the protocol and closed the substream(!). The maker however had not read `Msg3` from the stream yet. Since the substream was already closed the maker ran into an error because it tried to read from the closed substream.

I think this is a reasonable explanation. Consequences:

- I suppose this can happen more often.
- How do we ensure that the other party does not close a substream we are still reading from?

I think the solution is, that we should not use a pattern where both parties send and then wait to receive in parallel.
Instead of 

```
Party 1: send
Party 2: send
Party 1: wait for receive
Party 2: wait for receive
```

this should be:

```
Party 1: send
Party 2: wait for receive 
Party 1: wait for receive
Party 2: send
```

The old rollover protocol is bad protocol design 😅
I will think about how to best fix this.

**Comment by klochowicz** [comment]
Hm, this panic has happened 4 times in the last 24h. I am not sure whether we reached the bottom of this:
At least one of the panics was directly prepended with the new "FailureCounter" feature: https://itchysats.grafana.net/d/RvjotM97z/container-logs?orgId=1&var-setup=testnet&var-level=WARN&var-filter=&var-container=maker&from=1654725180000&to=1654725289000

**Comment by klochowicz** [comment]
> Hm, this panic has happened 4 times in the last 24h. I am not sure whether we reached the bottom of this:
> At least one of the panics was directly prepended with the new "FailureCounter" feature: https://itchysats.grafana.net/d/RvjotM97z/container-logs?orgId=1&var-setup=testnet&var-level=WARN&var-filter=&var-container=maker&from=1654725180000&to=1654725289000

update: it seems like it's the only one that coincided with failing the counter. Overall, there was 40 instances of dropping the connection to the counter triggering, which is worrisome. I would suggest enabling ping on the taker side.

**Comment by klochowicz** [comment]
Looking at the docs, we might be misusing the API: 
https://docs.rs/futures/0.3.14/futures/stream/trait.StreamExt.html#method.select_next_some

> If the future is polled after this [Stream](https://docs.rs/futures/0.3.14/futures/stream/trait.Stream.html) is empty it will panic. Using the future with a [FusedFuture](https://docs.rs/futures/0.3.14/futures/future/trait.FusedFuture.html)-aware primitive like the [select!](https://docs.rs/futures/0.3.14/futures/macro.select.html) macro will prevent this.

**Comment by da-kami** [comment]
Fixes:

https://github.com/itchysats/itchysats/pull/2207
https://github.com/itchysats/itchysats/pull/2211
https://github.com/itchysats/itchysats/pull/2217
https://github.com/itchysats/itchysats/pull/2220

**Comment by da-kami** [comment]
Unfortunately I have to re-open this ticket and go back to my original analysis and conclusion.
Yes, we found problems that were fixed, but we still see `panics` related to the taker closing the substream early.
We have another case where we run into the `Counterparty sent invalid revocation sk` problem.

This is again related to rollover panics on the `testnet` maker side.

Analysis by [query](https://itchysats.grafana.net/explore?orgId=1&left=%7B%22datasource%22:%22grafanacloud-itchysats-logs%22,%22queries%22:%5B%7B%22refId%22:%22Taker%22,%22expr%22:%22%7Bcontainer%3D%5C%22taker_master%5C%22,setup%3D%5C%22testnet%5C%22%7D%20%7C%3D%5C%22276cd584-673d-406b-8764-c0f54478f6df%5C%22%20!%3D%5C%22Applying%20new%20events%20to%20CFD%5C%22%20!%3D%5C%22collab_settlement%5C%22%20!%3D%5C%22settlement%5C%22%20!%3D%5C%22Settlement%5C%22%20!%3D%5C%22settle%5C%22%22,%22hide%22:false%7D,%7B%22refId%22:%22Maker%22,%22expr%22:%22%7Bcontainer%3D%5C%22maker%5C%22,setup%3D%5C%22testnet%5C%22%7D%20%7C%3D%5C%22276cd584-673d-406b-8764-c0f54478f6df%5C%22%20!%3D%5C%22Applying%20new%20events%20to%20CFD%5C%22%20!%3D%5C%22collab_settlement%5C%22%20!%3D%5C%22settlement%5C%22%20!%3D%5C%22Settlement%5C%22%20!%3D%5C%22settle%5C%22%22,%22hide%22:false%7D,%7B%22refId%22:%22A%22,%22expr%22:%22%7Bcontainer%3D%5C%22maker%5C%22,setup%3D%5C%22testnet%5C%22%7D%20%7C%3D%5C%22panic%5C%22%22,%22hide%22:false%7D%5D,%22range%22:%7B%22from%22:%221655091899950%22,%22to%22:%221655100469321%22%7D%7D)

Note: Rejected Collab settlements that happened in between the rollovers are irrelevant and thus filtered out.

First relevant block that shows the `panic` behaviour:

```
2022-06-13 14:00:13	taker_master	
{"timestamp":"2022-06-13T04:00:13.663544572Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverStarted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}Show context
	
2022-06-13 14:00:13	maker	
{"timestamp":"2022-06-13T04:00:13.705236742Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverStarted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 14:00:13	maker	
{"timestamp":"2022-06-13T04:00:13.993082757Z","level":"DEBUG","fields":{"message":"Handled request","method":"POST","path":"/api/cfd/276cd584-673d-406b-8764-c0f54478f6df/acceptRollover","status":"200 OK"},"target":"http"}
	
2022-06-13 14:00:14	maker	
{"timestamp":"2022-06-13T04:00:14.004895942Z","level":"DEBUG","fields":{"message":"Accepting rollover proposal","order_id":"276cd584-673d-406b-8764-c0f54478f6df","rollover_version":"V3","hours_to_charge":"2","funding_fee":"-0.00000003 BTC"},"target":"model::cfd"}
	
2022-06-13 14:00:14	maker	
{"timestamp":"2022-06-13T04:00:14.009053171Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverAccepted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 14:00:14	taker_master	
{"timestamp":"2022-06-13T04:00:14.152993238Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverAccepted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 14:00:14	taker_master	
{"timestamp":"2022-06-13T04:00:14.153630959Z","level":"ERROR","fields":{"message":"Rollover failed: Failed to get announcement: Announcement /x/BitMEX/BXBT/2022-06-14T05:00:00.price?n=20 not found","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"daemon::rollover::protocol"}
	
2022-06-13 14:00:14	taker_master	
{"timestamp":"2022-06-13T04:00:14.154288462Z","level":"WARN","fields":{"message":"Rollover failed: Failed to get announcement: Announcement /x/BitMEX/BXBT/2022-06-14T05:00:00.price?n=20 not found","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"model::cfd"}
	
2022-06-13 14:00:14	taker_master	
{"timestamp":"2022-06-13T04:00:14.263003333Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverFailed","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 14:00:14	maker	
thread 'rocket-worker-thread' panicked at 'SelectNextSome polled after terminated', /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/stream/stream/select_next_some.rs:32:9
```

The rollover failed early on the taker side because the taker was unable to fetch an announcement (this can happen and is not supposed to be a big deal, should just be retried).

Due to the taker finishing early the maker panicked when reading from the substream. This should not happen! If the substream gets closed by the taker the maker should finish gracefully and append failed. 
Due to the panic the maker never appended a `RolloverFailed` failed event.

We did not run into the `Counterparty sent invalid revocation sk` problem yet, but this showcases that the maker just panics when reading from the close stream.

The next rollover failed because the maker was in a state where it was "already rolling over" due to the panic:

```
2022-06-13 14:05:13	taker_master	
{"timestamp":"2022-06-13T04:05:13.644657412Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverStarted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 14:05:13	maker	
{"timestamp":"2022-06-13T04:05:13.646097071Z","level":"WARN","fields":{"message":"Rollover failed after handling taker proposal: Failed to execute command on CFD: The CFD is already being rolled over","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"daemon::rollover::maker"}
	
2022-06-13 14:05:13	maker	
{"timestamp":"2022-06-13T04:05:13.646141068Z","level":"ERROR","fields":{"message":"Rollover failed: Failed to execute command on CFD: The CFD is already being rolled over","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"daemon::rollover::protocol"}
	
2022-06-13 14:05:13	maker	
{"timestamp":"2022-06-13T04:05:13.647184839Z","level":"WARN","fields":{"message":"Rollover failed: Failed to execute command on CFD: The CFD is already being rolled over","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"model::cfd"}
	
2022-06-13 14:05:13	maker	
{"timestamp":"2022-06-13T04:05:13.649526189Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverFailed","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 14:05:14	taker_master	
{"timestamp":"2022-06-13T04:05:14.43053672Z","level":"ERROR","fields":{"message":"Rollover failed: End of stream while receiving rollover decision from maker","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"daemon::rollover::protocol"}
	
2022-06-13 14:05:14	taker_master	
{"timestamp":"2022-06-13T04:05:14.432997126Z","level":"WARN","fields":{"message":"Rollover failed: End of stream while receiving rollover decision from maker","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"model::cfd"}
	
2022-06-13 14:05:14	taker_master	
{"timestamp":"2022-06-13T04:05:14.439514732Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverFailed","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
```

We recover from this by appending `RolloverFailed` on both sides.

The next rollover ran into the problem originally described in this ticket:

```
2022-06-13 14:10:13	taker_master	
{"timestamp":"2022-06-13T04:10:13.697625011Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverStarted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 14:10:13	maker	
{"timestamp":"2022-06-13T04:10:13.704344886Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverStarted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 14:10:13	maker	
{"timestamp":"2022-06-13T04:10:13.844186027Z","level":"DEBUG","fields":{"message":"Handled request","method":"POST","path":"/api/cfd/276cd584-673d-406b-8764-c0f54478f6df/acceptRollover","status":"200 OK"},"target":"http"}
	
2022-06-13 14:10:13	maker	
{"timestamp":"2022-06-13T04:10:13.846271112Z","level":"DEBUG","fields":{"message":"Accepting rollover proposal","order_id":"276cd584-673d-406b-8764-c0f54478f6df","rollover_version":"V3","hours_to_charge":"2","funding_fee":"-0.00000022 BTC"},"target":"model::cfd"}Show context
	
2022-06-13 14:10:13	maker	
{"timestamp":"2022-06-13T04:10:13.850825082Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverAccepted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 14:10:13	taker_master	
{"timestamp":"2022-06-13T04:10:13.857140144Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverAccepted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 14:10:13	taker_master	
{"timestamp":"2022-06-13T04:10:13.85751487Z","level":"INFO","fields":{"message":"Rollover proposal got accepted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"daemon::rollover::taker"}
	
2022-06-13 14:10:29	taker_master	
{"timestamp":"2022-06-13T04:10:29.494185031Z","level":"INFO","fields":{"message":"Rollover was completed","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"model::cfd"}
	
2022-06-13 14:10:29	maker	
thread 'rocket-worker-thread' panicked at 'SelectNextSome polled after terminated', /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/stream/stream/select_next_some.rs:32:9
	
2022-06-13 14:10:29	taker_master	
{"timestamp":"2022-06-13T04:10:29.641485783Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverCompleted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 14:10:29	taker_master	
{"timestamp":"2022-06-13T04:10:29.64207173Z","level":"INFO","fields":{"message":"Rollover completed","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"daemon::rollover::protocol"}
```

The taker completed, the maker ran into the problem where the substream was already closed.

Interestingly we ran into another `panic` after this, the next rollover failed:

```
2022-06-13 16:00:17	taker_master	
{"timestamp":"2022-06-13T06:00:17.529061976Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverStarted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 16:00:17	maker	
{"timestamp":"2022-06-13T06:00:17.558389976Z","level":"WARN","fields":{"message":"Rollover failed after handling taker proposal: Failed to execute command on CFD: The CFD is already being rolled over","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"daemon::rollover::maker"}
	
2022-06-13 16:00:17	maker	
{"timestamp":"2022-06-13T06:00:17.558762401Z","level":"ERROR","fields":{"message":"Rollover failed: Failed to execute command on CFD: The CFD is already being rolled over","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"daemon::rollover::protocol"}Show context
	
2022-06-13 16:00:17	maker	
{"timestamp":"2022-06-13T06:00:17.564868219Z","level":"WARN","fields":{"message":"Rollover failed: Failed to execute command on CFD: The CFD is already being rolled over","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"model::cfd"}
	
2022-06-13 16:00:17	maker	
{"timestamp":"2022-06-13T06:00:17.56785069Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverFailed","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 16:00:17	taker_master	
{"timestamp":"2022-06-13T06:00:17.778386146Z","level":"ERROR","fields":{"message":"Rollover failed: End of stream while receiving rollover decision from maker","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"daemon::rollover::protocol"}
	
2022-06-13 16:00:17	taker_master	
{"timestamp":"2022-06-13T06:00:17.790046268Z","level":"WARN","fields":{"message":"Rollover failed: End of stream while receiving rollover decision from maker","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"model::cfd"}
	
2022-06-13 16:00:17	taker_master	
{"timestamp":"2022-06-13T06:00:17.798557126Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverFailed","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 16:00:19	maker	
thread 'rocket-worker-thread' panicked at 'SelectNextSome polled after terminated', /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/stream/stream/select_next_some.rs:32:9
```

The maker reported (again) that there was still an ongoing rollover (due to not appending failed because of the panic). The maker appended `RolloverFailed` then - but somehow there was still a task on the maker that panicked at that time! 

This would mean that the protocol started even though we already had appended failed? This should not happen.

From then onwards both parties run into the problem that the keys don't match because the taker is one rollover ahead:

```
2022-06-13 16:05:17	taker_master	
{"timestamp":"2022-06-13T06:05:17.590926364Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverStarted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 16:05:17	maker	
{"timestamp":"2022-06-13T06:05:17.819371489Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverStarted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 16:05:18	maker	
{"timestamp":"2022-06-13T06:05:18.01947613Z","level":"DEBUG","fields":{"message":"Handled request","method":"POST","path":"/api/cfd/276cd584-673d-406b-8764-c0f54478f6df/acceptRollover","status":"200 OK"},"target":"http"}
	
2022-06-13 16:05:18	maker	
{"timestamp":"2022-06-13T06:05:18.021206051Z","level":"DEBUG","fields":{"message":"Accepting rollover proposal","order_id":"276cd584-673d-406b-8764-c0f54478f6df","rollover_version":"V3","hours_to_charge":"4","funding_fee":"-0.00000043 BTC"},"target":"model::cfd"}
	
2022-06-13 16:05:18	maker	
{"timestamp":"2022-06-13T06:05:18.024230191Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverAccepted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 16:05:18	taker_master	
{"timestamp":"2022-06-13T06:05:18.031982164Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverAccepted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 16:05:18	taker_master	
{"timestamp":"2022-06-13T06:05:18.032208109Z","level":"INFO","fields":{"message":"Rollover proposal got accepted","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"daemon::rollover::taker"}
	
2022-06-13 16:05:26	taker_master	
{"timestamp":"2022-06-13T06:05:26.684788568Z","level":"ERROR","fields":{"message":"Rollover failed: Counterparty sent invalid revocation sk","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"daemon::rollover::protocol"}
	
2022-06-13 16:05:26	maker	
{"timestamp":"2022-06-13T06:05:26.68761941Z","level":"ERROR","fields":{"message":"Rollover failed: Counterparty sent invalid revocation sk","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"daemon::rollover::protocol"}
	
2022-06-13 16:05:26	maker	
{"timestamp":"2022-06-13T06:05:26.689228929Z","level":"WARN","fields":{"message":"Rollover failed: Counterparty sent invalid revocation sk","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"model::cfd"}
	
2022-06-13 16:05:26	taker_master	
{"timestamp":"2022-06-13T06:05:26.691372958Z","level":"WARN","fields":{"message":"Rollover failed: Counterparty sent invalid revocation sk","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"model::cfd"}
	
2022-06-13 16:05:26	maker	
{"timestamp":"2022-06-13T06:05:26.694353473Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverFailed","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
	
2022-06-13 16:05:26	taker_master	
{"timestamp":"2022-06-13T06:05:26.694483932Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverFailed","order_id":"276cd584-673d-406b-8764-c0f54478f6df"},"target":"sqlite_db"}
```

**Comment by da-kami** [comment]
TODO:

@klochowicz has pointed out that this problem might be related to how we use `select_next_some`.
The documentation states 

> If the future is polled after this [Stream](https://docs.rs/futures/0.3.14/futures/stream/trait.Stream.html) is empty it will panic. Using the future with a [FusedFuture](https://docs.rs/futures/0.3.14/futures/future/trait.FusedFuture.html)-aware primitive like the [select!](https://docs.rs/futures/0.3.14/futures/macro.select.html) macro will prevent this.

I agree with @klochowicz that we should move away from using `select_next_some` because in our context of using substreams of the connection itself this can cause trouble. The maker should not panic at this point but, at the worst case, time out!

However, I think in addition we need to do https://github.com/itchysats/itchysats/pull/2200 to ensure that the maker receives the last message. Otherwise it **can** happen that the taker finishes but the maker does not!

**Comment by da-kami** [comment]
Unfortunately I have to re-open this ticket yet another time.

The solution proposed by #2200 unfortunately does not solve this problem.
Furthermore, we run into this problem more frequently than I had found in my initial analysis.
Sorry for that, I did not connect these errors in my brain.

The taker fails to read the decision of the maker if there is a reject:

```
2022-06-17 09:21:25	
{"timestamp":"2022-06-16T23:21:25.914723123Z","level":"ERROR","fields":{"message":"Rollover failed: End of stream while receiving rollover decision from maker","order_id":"61836dd6-e2b9-4b2d-b5e1-c9b7337e6bb1"},"target":"daemon::rollover::protocol"}
```


The taker fails to read `Msg3` at the end after the maker closed (interestingly we see this on the taker now because of #2200):

```
2022-06-17 09:06:25	
{"timestamp":"2022-06-16T23:06:25.984998916Z","level":"INFO","fields":{"message":"Appended event to database","event":"RolloverFailed","order_id":"61836dd6-e2b9-4b2d-b5e1-c9b7337e6bb1"},"target":"sqlite_db"}
	
2022-06-17 09:06:25	
{"timestamp":"2022-06-16T23:06:25.980949189Z","level":"WARN","fields":{"message":"Rollover failed: Empty stream instead of Msg3","order_id":"61836dd6-e2b9-4b2d-b5e1-c9b7337e6bb1"},"target":"model::cfd"}
	
2022-06-17 09:06:25	
{"timestamp":"2022-06-16T23:06:25.980619095Z","level":"DEBUG","fields":{"message":"Applying new events to CFD","order_id":"61836dd6-e2b9-4b2d-b5e1-c9b7337e6bb1","aggregate":"model::cfd::Cfd","cfd_version":"8","num_events":"1"},"target":"sqlite_db"}
	
2022-06-17 09:06:25	
{"timestamp":"2022-06-16T23:06:25.980110587Z","level":"ERROR","fields":{"message":"Rollover failed: Empty stream instead of Msg3","order_id":"61836dd6-e2b9-4b2d-b5e1-c9b7337e6bb1"},"target":"daemon::rollover::protocol"}
```

We sporadically see this problem when processing offers from the maker as well:

```
2022-06-17 11:38:55	
{"timestamp":"2022-06-17T01:38:55.870071528Z","level":"DEBUG","fields":{"message":"Failed to process maker offers: The stream has terminated.","peer":"12D3KooWEsK2X8Tp24XtyWh7DM65VfwXtNH2cmfs2JsWmkmwKbV1"},"target":"xtra_libp2p_offer::taker"}
	
2022-06-17 02:35:10	
{"timestamp":"2022-06-16T16:35:10.030441065Z","level":"DEBUG","fields":{"message":"Failed to process maker offers: The stream has terminated.","peer":"12D3KooWEsK2X8Tp24XtyWh7DM65VfwXtNH2cmfs2JsWmkmwKbV1"},"target":"xtra_libp2p_offer::taker"}
```

These problems all boil down to  one party closing the stream and the other party trying to read from it. It is irrelevant which party opened the stream

I had a look at the [documentation of tokio-codec `Framed`](https://docs.rs/tokio-util/latest/tokio_util/codec/struct.Framed.html#method.new):

> Note that, for some byte sources, the stream can be resumed after an EOF by reading from it, even after it has returned None. Repeated attempts to do so, without new data available, continue to return None without creating more (closing) frames.

Note that this is not the `Framed` implementation that we use, we use https://github.com/mxinden/asynchronous-codec but I wonder if we can expect similar behavior there.

If so, that would mean that it can happen that we can actually poll again after we receive `None` and we could retrieve the value. But how would we know when to stop? I feel this is a very odd "feature" of `Framed` - that is also super hard to reproduce or test (I was unable to write a test reproducing this behavior so far...).

I am not sure what to do with the above documented behavior - usually when a stream returns `None` it is exhausted meaning there are no more messages. How would we properly handle this behavior? Retry until we hit a timeout? 

I don't think this is how this was intended to work.

**Comment by da-kami** [comment]
Note: I found a `set_read_after_close` in the Yamux config, but but evaluated that it can't be the problem. 
docs: https://paritytech.github.io/substrate/master/yamux/struct.Config.html#method.set_read_after_close

It is used in yamux when reading after the stream is closed:

https://github.com/libp2p/rust-yamux/blob/master/src/connection/stream.rs#L229-L231

The default value is set to `true`, and we are using the default value (verified by printing the yamux config upon opening a substream); if `true` then the if block will not be entered.

**Comment by da-kami** [comment]
I had a look at the `Framed` implementation that we use from `asynchronous-codec`. 
When reading there are two code pathes that can lead to return `Ready(None)`.
They are relevant when the read `buffer` is empty (my understanding of the code):

https://github.com/mxinden/asynchronous-codec/blob/master/src/framed_read.rs#L194-L209

I would assume that we should not end up in `ended` here because the read `buffer` should still not be `0`, i.e. we should not hit this code path but I am not 100% sure.

**Comment by da-kami** [comment]
I had a look at how we implement `AsyncRead` on the substream:

https://github.com/itchysats/itchysats/blob/415ecad93230a59cf0e28286d28893dfc4cda18b/xtra-libp2p/src/substream.rs#L70-L96

I don't see how this could be the problem yet but I don't fully understand the `?`ed `poll_read` with `ready!` macro yet.

The underlying `poll_read` should be yamux:

https://github.com/libp2p/rust-yamux/blob/master/src/connection/stream.rs#L277-L326

**Comment by klochowicz** [comment]
@luckysori investigated these issues, and after having a discussion we decided to open separate issues to discuss the way forward. Both issues happened due to underlying protocol changes (new recent invariants / checks), not libp2p. 

#2259 
#2255 

Please feel free to reopen the issue if the tickets above do not exhaust the issues mentioned in this ticket.

**Comment by luckysori** [comment]
We were also able to identify another issue related to timing out when sending a collab-settlement or rollover proposal because the maker wasn't able to process the `NewInboundSubstream` messages in time: #2260. 

It is also hard to worry too much about the new network stack based on the problems outlined in #2260 either because:

- These problems are infrequent on `testnet`. Over the past 30 days we've had one instance of the collab-settlement error and 16 of the rollover one (13 of those concentrated over 6 hours).

Nevertheless, we should investigate and mitigate them because we want a more stable application (both on `testnet` and `mainnet`).

**Comment by da-kami** [comment]
Sorry, but I don't think the new tickets cover the problems discussed in this ticket.

Given that the timestamps of the taker's completion and the maker's panic / `Empty stream` messages I don't think this is a problem where the taker times out. We had several rollovers where the taker succeeded (!) and the maker failed when reading from an empty stream.

**Comment by da-kami** [comment]
Feel free to use the query linked here to have a nice overview:

https://github.com/itchysats/itchysats/issues/2199#issuecomment-1153555870

Back then we ran into panics, but the problem with `Empty stream` is still the same. The maker (or taker) reads from a stream that is already terminated.

This also happens for the `offers` protocol as reported here:

https://github.com/itchysats/itchysats/issues/2199#issuecomment-1158334633

**Comment by klochowicz** [comment]
@da-kami would you mind creating *specific* tickets instead? this catch-all ticket is really unfortunate, because there's a high risk of conflating different issues into one. I would very much prefer closing the ticket (as the original issue mentioned and described has been reopened in different tickets). It would have been much simpler to tackle the issues if you've opened *new* tickets instead for specific problems.
I will seek to do that for you today, so we can actually action things.

Thank you for pointing out additional queries, so we can have a look at them today.

**Comment by da-kami** [comment]
This is not a "catching all" ticket. The problems are related and to my mind caused by the behavior that we try to read from an empty (i.e. "terminated") stream. I think this is the only problem, hence this ticket.

**Comment by da-kami** [comment]
> as the original issue mentioned and described has been reopened in different tickets

Sorry, but these tickets do not reflect the problem.

**Comment by klochowicz** [comment]
> Sorry, but I don't think the new tickets cover the problems discussed in this ticket.
> 
> Given that the timestamps of the taker's completion and the maker's panic / `Empty stream` messages I don't think this is a problem where the taker times out. We had several rollovers where the taker succeeded (!) and the maker failed when reading from an empty stream.

please reference this in a new issue (if possible, with the logs to grafana so we can find that).  this particular information was not present in this ticket.

**Comment by da-kami** [comment]
Superseded by https://github.com/itchysats/itchysats/issues/2262

**Comment by da-kami** [comment]
> this particular information was not present in this ticket.

It was mentioned, but granted you would have had to take a look at the logs I showed in my comments. I hope the new ticket that provides all the queries ready in links helps

**Comment by klochowicz** [comment]
> > this particular information was not present in this ticket.
> 
> It was mentioned, but granted you would have had to take a look at the logs I showed in my comments. I hope the new ticket that provides all the queries ready in links helps

it is helpful, thanks for taking time to reopen more focused ticket 👍
